### ISSUE TITLE
Improve OAI-PMH set spec compliance

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#4235]
OAI sets, [per the OAI spec](https://www.openarchives.org/OAI/2.0/openarchivesprotocol.htm#Set), should be specified using colon-separated elements consisting of

> a string consisting of any valid [URI unreserved characters](http://www.ietf.org/rfc/rfc2396.txt)

...which in return consist of...

```
2.3. Unreserved Characters

   Data characters that are allowed in a URI but do not have a reserved
   purpose are called unreserved.  These include upper and lower case
   letters, decimal digits, and a limited set of punctuation marks and
   symbols.

      unreserved  = alphanum | mark

      mark        = "-" | "_" | "." | "!" | "~" | "*" | "'" | "(" | ")"

   Unreserved characters can be escaped without changing the semantics
   of the URI, but this should not be done unless the URI is being used
   in a context that does not allow the unescaped character to appear.
```
However, our current code simply `urlencode`s the elements, which results in them containing `%`-escaped sequences, `+` characters for spaces, etc. That breaks compliance with the specification.

Improve the compliance with the specification by ensuring that set specifiers don't include forbidden characters, while still making sure that they're sufficiently unique.

(The changes need to be made e.g. in `OAIDAO.inc.php`.)

**Comment by isgrim** [comment]
Dear Alec @asmecher ,

I stumbled upon this issue because the German National Lilbrary is using our OJS OAI Interface for harvesting URNs for registration, we use an OAI Metadataformat Plugin for this, which was orignally developed by Bozana  (our fork here: https://github.com/UB-Heidelberg/epicur/) and adjusted for OJS3 by Phillip Franck from DAI.

The national library now uses a new harvester, which strictly validates the OAI, and because of the "setSpec" Field in the OAI header the XML is considered invalid.

The changes can go very deep, because of the data object tombstones. I tried to find all the places where the setSpec is formatted/parsed, but its all over the place. I would like to fix this myself, but I have too many undone work items here.

Do you know anyone who could jump in?

Regards,

Nils

**Comment by asmecher** [comment]
Hi @isgrim -- I can maybe help out here. The set spec is only formatted in a few places (decoded in `classes/oai/ojs/JournalOAI.inc.php` in `setSpecToSectionId` and encoded in `classes/oai/ojs/OAIDAO.inc.php`) and it's only stored in the tombstone DAO.

So correcting this would involve...
- Coming up with a good alternative to the `urlencode`/`urldecode` functions for encoding (I would suggest putting these in `lib/pkp/classes/oai/OAIUtils.inc.php`)
- Replacing `urlencode`/`urldecode` calls in the above classes with these functions
- Writing an upgrade script to convert existing data in the tombstones table.

Can we divide this up between the two of us? Is there part of it that you've already worked with?

**Comment by isgrim** [comment]
Hi Alec @asmecher ,

thanks for your answer! I think I could handle:

> Coming up with a good alternative to the urlencode/urldecode functions for encoding (I would suggest putting these in lib/pkp/classes/oai/OAIUtils.inc.php)

and

> Replacing urlencode/urldecode calls in the above classes with these functions

But I would need help with the upgrade script.

Thanks again for your kind offer!

Regards,

Nils

**Comment by asmecher** [comment]
:+1: :+1: Thanks, that's great, @isgrim!

**Comment by olovy** [comment]
Hi @asmecher and @isgrim, 

We stumbled upon the same issue. Have you made any progress on the fix?
We will probably try to fix it ourselves otherwise.

(At the National Library of Sweden we use our own OAI-PMH [metadata format plugin](https://github.com/Kungbib/fgs-oaipmh-ojs-plugin) when harvesting articles for long term storage from our new OJS-based platform [Publicera](https://publicera.kb.se/)).

Regards,
Olov

**Comment by asmecher** [comment]
I'm not aware of any further progress on this -- some assistance would be very welcome!

**Comment by isgrim** [comment]
Hi Olov @olovy and Alec @asmecher ,

I'm sorry for not going further with this. This is still relevant for us, but It was not high priority.

If you have the time and resources to contribute the fix. This would be very much appreciated.
I could work on this after my current task, maybe in one or two weeks.

Regards,

Nils

**Comment by olovy** [comment]
@asmecher, @isgrim 

We probably will have someone available to look at this in our next period, starting 2021-09-20.

**Comment by asmecher** [comment]
Thanks, @olovy! I reviewed both PRs; specifics are over there.

**Comment by asmecher** [comment]
Thanks very much, @olovy -- that's a nice, clean solution and it's great to have the patches for OMP and OPS as well. I tested these and made a few slight changes but nothing major :+1:

**Comment by olovy** [comment]
Thank you, @asmecher !

**Comment by nils-stefan-weiher** [comment]
Hi Alec @asmecher ,

do you know if it would be feasible to backport this fix for the OJS 3.2 branch?

We might upgrade to OJS 3.4 only next year, but need this fix for our production setup.

Do you think it would be possible at all?

Best regards,

Nils Weiher

**Comment by asmecher** [comment]
@nils-stefan-weiher, we aren't developing the 3.2 branch anymore, and don't plan any more releases from it; if you're interesting in backporting it yourself I could look over a PR, though. I might suggest looking at upgrading to the LTS release, though -- you'll get this fix, plus a stable platform that has a longer window of maintenance and is available now and stable!

**Comment by nils-stefan-weiher** [comment]
Oh I thought this fix will be only available for 3.4?

I will propably adopt this fix for OJS 3.2, but I dont know if I have the time to make a proper PR. Will let you know.

**Comment by asmecher** [comment]
@nils-stefan-weiher, whoops, you're absolutely correct -- this isn't included in the `stable-3_3_0` branch. A back-port would be a little invasive and include a database update, which is why it wasn't back-ported to `stable-3_3_0` in the first place. You might need to keep it as a local adaptation if you do apply it back to an earlier release.
